File: src/container/fs.rs
use std::fs;
use std::path::Path;
use std::process::Command;
use nix::mount::{mount, MsFlags, umount2, MntFlags};
use crate::config::QUBE_CONTAINERS_BASE;

pub fn mount_volume(cid: &str, host_path: &str, container_path: &str) -> Result<(), std::io::Error> {
    if !Path::new(host_path).exists() {
        return Err(std::io::Error::new(std::io::ErrorKind::NotFound,
            format!("Host path '{}' does not exist", host_path)));
    }

    let rootfs = get_rootfs(cid);
    let dest = format!("{}/{}", rootfs, container_path.trim_start_matches('/'));

    if !Path::new(&dest).exists() {
        println!("DEBUG: Creating mount point at {}", dest);
        fs::create_dir_all(&dest)?;
    }    

    println!("DEBUG: Attempting to mount {} -> {}", host_path, dest);

    let _ = umount2(Path::new(&dest), MntFlags::MNT_DETACH);

    let rootfs_path = std::path::Path::new(&rootfs);
    let _ = mount(
        Some(rootfs_path),
        rootfs_path,
        None::<&str>,
        MsFlags::MS_REC | MsFlags::MS_SHARED,
        None::<&str>,
    );
    Ok(())
}

pub fn get_rootfs(cid: &str) -> String {
    format!("{}/rootfs", format!("{}/{}", QUBE_CONTAINERS_BASE, cid))
}

pub fn prepare_rootfs_dir(cid: &str) {
    let rootfs = get_rootfs(cid);
    if Path::new(&rootfs).exists() {
        fs::remove_dir_all(&rootfs).ok();
    }
    fs::create_dir_all(&rootfs).unwrap();
}

pub fn copy_directory_into_home(cid: &str, work_dir: &str) {
    let home_path = format!("{}/home", get_rootfs(cid));
    if !Path::new(&home_path).exists() {
        fs::create_dir_all(&home_path).ok();
    }
    let status = Command::new("cp")
        .args(["-r", &format!("{}/.", work_dir), &home_path])
        .status()
        .unwrap();
    if !status.success() {
        eprintln!("Warning: copying {} -> {} failed.", work_dir, home_path);
    }
}

pub fn mount_proc(cid: &str) -> Result<(), std::io::Error> {
    use nix::mount::{mount, MsFlags};
    let proc_path = format!("{}/proc", get_rootfs(cid));
    fs::create_dir_all(&proc_path)?;
    mount(
        Some("proc"),
        proc_path.as_str(),
        Some("proc"),
        MsFlags::MS_NOEXEC | MsFlags::MS_NOSUID | MsFlags::MS_NODEV,
        None::<&str>,
    )
    .map_err(|e| std::io::Error::new(std::io::ErrorKind::Other, e))
}

File: internal/core/container/fs.go
package container

import (
	"fmt"
	"io"
	"io/ioutil"
	"os"
	"os/exec"
	"path/filepath"
	"strings"

	"github.com/Voyrox/Qube/internal/config"
	"golang.org/x/sys/unix"
)

func GetRootfs(cid string) string {
	return filepath.Join(config.QubeContainersBase, cid, "rootfs")
}

func PrepareRootfsDir(cid string) error {
	rootfs := GetRootfs(cid)
	if _, err := os.Stat(rootfs); err == nil {
		if err := os.RemoveAll(rootfs); err != nil {
			return err
		}
	}
	return os.MkdirAll(rootfs, 0755)
}

func CopyDirectoryIntoHome(cid, workDir string) error {
	rootfs := GetRootfs(cid)
	workspacePath := filepath.Join(rootfs, "workspace")

	if err := os.MkdirAll(workspacePath, 0755); err != nil {
		return err
	}

	rsyncCmd := exec.Command("rsync", "-a", "--exclude=.git", workDir+"/", workspacePath+"/")
	if err := rsyncCmd.Run(); err == nil {
		return nil
	}

	cpCmd := exec.Command("sh", "-c", fmt.Sprintf("cp -rT %s %s", workDir, workspacePath))
	if err := cpCmd.Run(); err != nil {
		return fmt.Errorf("failed to copy %s -> %s: %w", workDir, workspacePath, err)
	}

	return nil
}

func MountProc(cid string) error {
	rootfs := GetRootfs(cid)
	procPath := filepath.Join(rootfs, "proc")

	if err := os.MkdirAll(procPath, 0755); err != nil {
		return err
	}

	return unix.Mount("proc", procPath, "proc", unix.MS_NOEXEC|unix.MS_NOSUID|unix.MS_NODEV, "")
}

func MountVolume(cid, hostPath, containerPath string) error {
	if _, err := os.Stat(hostPath); os.IsNotExist(err) {
		return fmt.Errorf("host path '%s' does not exist", hostPath)
	}

	rootfs := GetRootfs(cid)
	dest := filepath.Join(rootfs, strings.TrimPrefix(containerPath, "/"))

	if _, err := os.Stat(dest); os.IsNotExist(err) {
		fmt.Printf("DEBUG: Creating mount point at %s\n", dest)
		if err := os.MkdirAll(dest, 0755); err != nil {
			return err
		}
	}

	fmt.Printf("DEBUG: Attempting to mount %s -> %s\n", hostPath, dest)

	unix.Unmount(dest, unix.MNT_DETACH)

	return unix.Mount(hostPath, dest, "", unix.MS_BIND, "")
}

func CopyFile(src, dst string) error {
	sourceFile, err := os.Open(src)
	if err != nil {
		return err
	}
	defer sourceFile.Close()

	destFile, err := os.Create(dst)
	if err != nil {
		return err
	}
	defer destFile.Close()

	_, err = io.Copy(destFile, sourceFile)
	return err
}

func CopyDir(src, dst string) error {
	entries, err := ioutil.ReadDir(src)
	if err != nil {
		return err
	}

	if err := os.MkdirAll(dst, 0755); err != nil {
		return err
	}

	for _, entry := range entries {
		srcPath := filepath.Join(src, entry.Name())
		dstPath := filepath.Join(dst, entry.Name())

		if entry.IsDir() {
			if err := CopyDir(srcPath, dstPath); err != nil {
				return err
			}
		} else {
			if err := CopyFile(srcPath, dstPath); err != nil {
				return err
			}
		}
	}

	return nil
}
